### PR TITLE
LPS-180298 POC: Manage multiple store areas in DLStore

### PR DESCRIPTION
**Pull request by marcogalluzzi**
This is a proposal for the implementation of multiple store areas (I think the term _area_ is better than _space_). The first idea was to set thread locally the read and write areas, to avoid doing many changes. However, the concept of an _area_ should be included in every class implementing the `Store` interface, otherwise we could not change, for instance, the `rootDir` of an `FileSystemStore`. So, if we have to finally change the stores, I would add the _area_ as a parameter for all `Store` methods. In order to avoid many changes at first, I've added an intermediate interface called `AreaStore` that is only implemented by the `FileSystemStore`. All stores that want to use the _areas_ will end implementing the `AreaStore`: if all, then we could simply end changing the `Store` interface and remove the `AreaStore` interface.

Finally, in class `DLStoreImpl` I've just implemented an example of how the deletion should be done using the new `StoreArea`, there are still operations missing.

### Open questions

`FileSystemStoreRegister.acivate()` and `AdvancedFileSystemStoreRegister.activate()` register the service with parameter `rootDir`. Currently we are passing only the `AreaType.LIVE` root folder. I think this parameter is only used by the `AntivirusAsyncFileStoreMessageListener` to do the Antivirus working on the root folder. Should we pass to the Antivirus also the root folders `AreaType.EVICTED` and `AreaType.DELETED` ?

I think we should change the service `CTSContentLocalService` updated in class `CTStore` to include the `areaType` or we have to update it only when `AreaType.LIVE` is touched ?

### CODE REVIEW

**Comment by liferay-lima** [comment]
ci:test:sf

**Comment by liferay-lima** [comment]
ci:test:relevant

**Review by adolfopa — CHANGES_REQUESTED** [review]
@marcogalluzzi I'm not sure about this, this is a very big API change. As this affects an component that we don't own we'll need to speak with core infrastructure, show them both approaches and let them decide.

File: modules/apps/portal-store/portal-store-file-system/src/main/java/com/liferay/portal/store/file/system/FileSystemStore.java [hunk 1]
```diff
@@ -218,36 +232,37 @@ public String[] getFileVersions(
 		return versions;
 	}
 
-	public File getRootDir() {
-		return _rootDir;
+	public File getRootDir(AreaType areaType) {
+		return _rootDirs.get(areaType);
```
[the remark this answers is not in the document]
Comment (adolfopa) [reply modules/apps/portal-store/portal-store-file-system/src/main/java/com/liferay/portal/store/file/system/FileSystemStore.java:236]: Another thing to take into account is that the root dir is an implementation detail of the filesystem store. In other stores there is no concept of root dir. The current approach for the AreaType may not work for other stores.

**Pull request closed without merging**